lib/netext/grpcext/conn.go

```diff
@@ -0,0 +1,313 @@
+// Package grpcext allows gRPC requests collecting stats info.
+package grpcext
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+	"go.k6.io/k6/js/modules"
+	"go.k6.io/k6/metrics"
+
+	protov1 "github.com/golang/protobuf/proto" //nolint:staticcheck,nolintlint // this is the old v1 version
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	grpcstats "google.golang.org/grpc/stats"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/encoding/prototext"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/dynamicpb"
+)
+
+// Request represents a gRPC request.
+type Request struct {
+	MethodDescriptor protoreflect.MethodDescriptor
+	Tags             map[string]string
+	Message          []byte
+}
+
+// Response represents a gRPC response.
+type Response struct {
+	Message  interface{}
+	Error    interface{}
+	Headers  map[string][]string
+	Trailers map[string][]string
+	Status   codes.Code
+}
+
+type clientConnCloser interface {
+	grpc.ClientConnInterface
+	Close() error
+}
+
+// Conn is a gRPC client connection.
+type Conn struct {
+	raw clientConnCloser
+}
+
+// DefaultOptions generates an option set
+// with common options for requests from a VU.
+func DefaultOptions(vu modules.VU) []grpc.DialOption {
+	dialer := func(ctx context.Context, addr string) (net.Conn, error) {
+		return vu.State().Dialer.DialContext(ctx, "tcp", addr)
+	}
+
+	return []grpc.DialOption{
+		grpc.WithBlock(),
+		grpc.FailOnNonTempDialError(true),
+		grpc.WithReturnConnectionError(),
+		grpc.WithStatsHandler(statsHandler{vu: vu}),
+		grpc.WithContextDialer(dialer),
+	}
+}
+
+// Dial establish a gRPC connection.
+func Dial(ctx context.Context, addr string, options ...grpc.DialOption) (*Conn, error) {
+	conn, err := grpc.DialContext(ctx, addr, options...)
+	if err != nil {
+		return nil, err
+	}
+	return &Conn{
+		raw: conn,
+	}, nil
+}
+
+// ReflectionClient returns a reflection client based on the current connection.
+func (c *Conn) ReflectionClient() (*ReflectionClient, error) {
+	return &ReflectionClient{Conn: c.raw}, nil
+}
+
+// Invoke executes a unary gRPC request.
+func (c *Conn) Invoke(
+	ctx context.Context,
+	method string,
+	md metadata.MD,
+	req Request,
+	opts ...grpc.CallOption) (*Response, error) {
+	if method == "" {
+		return nil, fmt.Errorf("method is required")
+	}
+	if req.MethodDescriptor == nil {
+		return nil, fmt.Errorf("request method descriptor is required")
+	}
+	if len(req.Message) == 0 {
```

Comment:
:thinking: should a message always be presented?

Reply:
Yes, at least a `{}` must be there. The change in https://github.com/grafana/k6/pull/2484/commits/8b64a1414a1939fceef051447911c21f039831bf should make it visible. This is related to the `protojson` dependency, where a valid JSON must be used.